File: app/src/test/java/com/github/ivanshafran/unit_tests_evolution/v3/ShowRateUsLogicTest.java
package com.github.ivanshafran.unit_tests_evolution.v3;

import android.content.Context;
import com.github.ivanshafran.sharedpreferencesmock.SPMockBuilder;
import com.github.ivanshafran.unit_tests_evolution.*;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.util.Calendar;
import java.util.concurrent.TimeUnit;

public class ShowRateUsLogicTest {

    private RateUsPreferences rateUsPreferences;
    private BuyPreferences buyPreferences;
    private Time time;

    private ShowRateUsLogic showRateUsLogic;

    @Before
    public void setUp() {
        final Context mockedContext = new SPMockBuilder().createContext();
        rateUsPreferences = new RateUsPreferencesImpl(mockedContext);
        buyPreferences = new BuyPreferencesImpl(mockedContext);
        time = Mockito.mock(Time.class);

        showRateUsLogic = new ShowRateUsLogic(
                rateUsPreferences,
                buyPreferences,
                time
        );
    }

    @Test
    public void onFirstCheckAndOneClickItShouldNotShow() {
        buyPreferences.incrementBuyCount();

        Assert.assertFalse(showRateUsLogic.shouldShowRateUs());
    }

    @Test
    public void onThreeClicksAndItShouldShow() {
        // clicked three times
        buyPreferences.incrementBuyCount();
        buyPreferences.incrementBuyCount();
        buyPreferences.incrementBuyCount();

        // set first dialog show time
        final Calendar calendar = Calendar.getInstance();
        calendar.set(2019, Calendar.JULY, 7);
        rateUsPreferences.setLastShownTimeMillis(calendar.getTimeInMillis());

        // set current time to be 90 days after first show
        Mockito.when(time.getCurrentTimeMillis()).thenReturn(calendar.getTimeInMillis() + TimeUnit.DAYS.toMillis(90));

        Assert.assertTrue(showRateUsLogic.shouldShowRateUs());
    }
}

File: app/src/test/java/com/github/ivanshafran/unit_tests_evolution/v5/ShowRateUsLogicTest.kt
package com.github.ivanshafran.unit_tests_evolution.v5

import com.github.ivanshafran.sharedpreferencesmock.SPMockBuilder
import com.github.ivanshafran.unit_tests_evolution.BuyPreferencesImpl
import com.github.ivanshafran.unit_tests_evolution.RateUsPreferencesImpl
import com.github.ivanshafran.unit_tests_evolution.ShowRateUsLogic
import com.github.ivanshafran.unit_tests_evolution.Time
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import java.util.*
import java.util.concurrent.TimeUnit

class ShowRateUsLogicTest : Spek({

    val calendar = Calendar.getInstance()
    calendar.set(2019, Calendar.JULY, 7)

    val SOME_DAY_IN_MILLIS = calendar.timeInMillis
    val LESS_THAN_TWO_MONTHS = TimeUnit.DAYS.toMillis(59)
    val MORE_THAN_TWO_MONTHS = TimeUnit.DAYS.toMillis(61)

    val mockedContext = SPMockBuilder().createContext()

    val rateUsPreferences by memoized { RateUsPreferencesImpl(mockedContext) }
    val buyPreferences by memoized { BuyPreferencesImpl(mockedContext) }
    val time by memoized { mock<Time>() }
    val showRateUsLogic by memoized {
        ShowRateUsLogic(
            rateUsPreferences,
            buyPreferences,
            time
        )
    }

    fun prepareConditions(
        buyClickedTimes: Int = 0,
        isNeverShownAgainClicked: Boolean = false,
        isRateNowClicked: Boolean = false,
        lastShownTimeMillis: Long = 0,
        currentTimeMillis: Long = System.currentTimeMillis()
    ) {
        repeat(buyClickedTimes) { buyPreferences.incrementBuyCount() }

        if (isNeverShownAgainClicked) rateUsPreferences.setNeverShownAgainClicked()
        if (isRateNowClicked) rateUsPreferences.setRateNowClickedClicked()
        rateUsPreferences.setLastShownTimeMillis(lastShownTimeMillis)

        whenever(time.getCurrentTimeMillis()).thenReturn(currentTimeMillis)
    }

    describe("show rate us logic") {

        context("first conditions checks") {
            context("buy clicked once") {
                beforeEachTest {
                    prepareConditions(buyClickedTimes = 1)
                }

                it("should not show 'rate us'") {
                    Assert.assertFalse(showRateUsLogic.shouldShowRateUs())
                }
            }

            context("buy clicked two times") {
                beforeEachTest {
                    prepareConditions(buyClickedTimes = 2)
                }

                it("should show 'rate us'") {
                    Assert.assertTrue(showRateUsLogic.shouldShowRateUs())
                }
            }
        }

        context("'rate us' was shown already") {
            context("user clicked 'show me later' on the dialog") {
                context("less than two months passed and user clicks buy") {
                    beforeEachTest {
                        prepareConditions(
                            buyClickedTimes = 3,
                            lastShownTimeMillis = SOME_DAY_IN_MILLIS,
                            currentTimeMillis = SOME_DAY_IN_MILLIS + MORE_THAN_TWO_MONTHS
                        )
                    }

                    it("should show 'rate us' again") {
                        Assert.assertTrue(showRateUsLogic.shouldShowRateUs())
                    }
                }

                context("more than two months passed and user clicks buy") {
                    beforeEachTest {
                        prepareConditions(
                            buyClickedTimes = 3,
                            lastShownTimeMillis = SOME_DAY_IN_MILLIS,
                            currentTimeMillis = SOME_DAY_IN_MILLIS + MORE_THAN_TWO_MONTHS
                        )
                    }

                    it("should show 'rate us' again") {
                        Assert.assertTrue(showRateUsLogic.shouldShowRateUs())
                    }
                }
            }
        }
    }
})
